dali/bus.py

```diff
@@ -130,11 +128,93 @@ def assign_short_addresses(self):
                     i.send(gear.ProgramShortAddress(new_addr))
                     r = i.send(gear.VerifyShortAddress(new_addr))
                     if r.value is not True:
-                        raise ProgramShortAddressFailure(new_addr)
+                        ProgramShortAddressFailure(new_addr)
+                        print(f"Error in programming short address {new_addr}")
                     i.send(gear.Withdraw())
-                    Device(address=new_addr, bus=self)
+                    Device(address=new_addr, randomAddress=low, bus=self)
                 else:
                     i.send(gear.Terminate())
                     raise NoFreeAddress()
                 low = low + 1
         i.send(gear.Terminate())
+
+    def assign_short_addresses(self):
+        """Search for devices on the bus with no short address allocated, and
+        allocate each one a short address from the set of unused
+        addresses.
+        """
+        if not self._bus_scanned:
+            self.scan()
+        self.search_bus(broadcast=False)
+        self.query_device_types()
+        self.query_groups()
+
+    def initialize_bus(self):
+        """Initialize bus
+        """
+        self._devices = {}
+        self.search_bus(broadcast=True)
+        self.query_device_types()
+        self.query_groups()
+
+    def query_device_types(self):
+        """Find the device types of the devices in the bus
+        """
+        i = self.get_interface()
+        for sa in range(64):
+            if sa in self._devices:
+                self._devices[sa].deviceType = i.send(gear.QueryDeviceType(sa))
+
+    def query_groups(self):
+        """Find the groups of the devices in the bus
+        """
+        i = self.get_interface()
+        for sa in range(64):
+            if sa in self._devices:
+                group1 = i.send(gear.QueryGroupsZeroToSeven(sa))
+                group2 = i.send(gear.QueryGroupsEightToFifteen(sa))
+                try:
+                    group1 = group1.value.as_integer
+                    group2 = group2.value.as_integer
+                except:
+                    self._devices[sa].group = None
+                else:
+                    self._devices[sa].groups = self.parse_groups(group1, group2)
+
+    def parse_groups(self, group1, group2):
```

Comment:
This all function can be summarised as
```
return str([j for j in [group1 & 1 << i for i in range(7)] if j != 0])[1:-1] + str([j+8 for j in [group2 & 1 << i for i in range(7)] if j != 0])[1:-1]

Reply:
Might be. Not sure where the commas are added but need to test this. I'm not sure if this is the most self explaining way to do it.

Reply:
commas are added because we are casting the list into a str